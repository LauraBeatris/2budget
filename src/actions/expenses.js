import database from "../firebase/firebase";

//ADD_EXPENSE
export const addExpense = expense => {
  return {
    type: "ADD_EXPENSE",
    expense
  };
};

export const addExpenseRequest = (expenseData = {}) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    };
    // After sending this to database -> dispatch the action to the store
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update
});

//SET_EXPENSE
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const expenses = [];
    const { uid } = getState().auth;

    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};

export const startRemoveExpenses = ({ id } = {}) => {
  // This asyncronous function is responsable for doing a connection with
  // firebase and dispatching an action
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(snapshot => {
        //console.log("expense removed");
        dispatch(removeExpense({ id }));
      });
  };
};

export const startEditExpenses = (id, update) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    // Getting the reference of the expense from database and setting the new value
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(update)
      .then(snapshot => {
        //expense edit

        //Dispatching the action to store
        dispatch(editExpense(id, update));
      })
      .catch(err => console.log(err));
  };
};
