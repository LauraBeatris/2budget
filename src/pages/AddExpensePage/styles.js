import styled from "styled-components";

export const AddContainer = styled.div`
  text-align: center;

  .add-title {
    margin-top: 5rem;
    font-size: 2rem;
    font-weight: 100;
    color: rgb(77, 81, 94);
  }

  animation-name: goin;
  animation-duration: 0.5s;

  @keyframes goin {
    from {
      opacity: 0;
      transform: translateX(-10px);
    }

    to {
      opacity: 1;
      transform: translateX(0px);
    }
  }
`;
