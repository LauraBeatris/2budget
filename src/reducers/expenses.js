const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update // overriden the expense items with the action updates icons
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
