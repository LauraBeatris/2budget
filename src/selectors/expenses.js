import moment from "moment";

// Get visible expenses
export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return b.createdAt < a.createdAt ? -1 : 1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? -1 : 1;
      } else {
        return null;
      }
    });
};

export const getExpenseTotal = expenses => {
  if (expenses.length === 0) {
    return 0;
  } else if (expenses.length > 1) {
    const amounts = expenses.map(expense => expense.amount);
    return amounts.reduce((sum, value) => sum + value, 1);
  } else {
    return expenses[0].amount;
  }
};
