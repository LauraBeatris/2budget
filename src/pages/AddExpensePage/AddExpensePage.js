import React, {Component} from "react";
import ExpenseForm from "../../components/ExpenseForm/ExpenseForm";
import Header from "../../components/Header/Header";
import { connect } from "react-redux";
import { addExpense } from "../../actions/expenses";
import { AddContainer } from "./styles";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.onSubmit(expense)
    // Redirecting to dashboard page
    this.props.history.push("/dashboard");
  }
  render(){
    return (
       <div>
        <Header />
        <AddContainer>
          <h1 className="add-title">Add a expense</h1>
          <ExpenseForm
            // Dispatching the action ADD_EXPENSE to the redux store
            onSubmit={this.onSubmit}
          />
        </AddContainer>
      </div>
    )
   
  }
}

const mapDispatchToProps = (dispatch) => (
   {
    onSubmit: (expense) => dispatch(addExpense(expense))
  }
)

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
