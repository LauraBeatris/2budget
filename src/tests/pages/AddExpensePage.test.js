import React from "react";
import { shallow } from "enzyme";
import { AddExpensePage } from "../../pages/AddExpensePage/AddExpensePage";
import expenses from "../fixtures/expenses";

let addExpense, history, wrapper;

// Jest lifecycle method -> will run this code before each test
beforeEach(() => {
  // Setting values to the test spies
  history = {
    push: jest.fn()
  };
  addExpense = jest.fn();

  wrapper = shallow(
    <AddExpensePage addExpense={addExpense} history={history} />
  );
});

test("should render AddExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
  // Finding the element and calling the props with a expense
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);

  // Checking if they were called with the correctly data
  expect(history.push).toHaveBeenLastCalledWith("/dashboard");
  expect(addExpense).toHaveBeenLastCalledWith(expenses[0]);
});
