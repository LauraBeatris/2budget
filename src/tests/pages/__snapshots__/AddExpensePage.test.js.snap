// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
       2Budget | Add expense 
    </title>
  </HelmetWrapper>
  <.sc-bwzfXH>
    <h1
      className="add-title"
    >
      Add a expense
    </h1>
    <ExpenseForm
      onSubmit={[Function]}
      type="Add"
    />
  </.sc-bwzfXH>
</div>
`;
